### PR TITLE
feat(app, app-shell, api-client): include analysis as cli tool within app for protocol ingestion 

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

Expose the protocol analysis engine functionality from the opentrons python module. Add machinery to
allow this cli to be called from within the running Opentrons App via a configurable path to a
python interpretter. Wire up the protocol file analysis to the existing protocol file system storage
machinery. Wire up analysis data to protocols landing and details page scaffolding.

Feel free to take a look at the code and test things out.  For efficiency's sake I will be adding unit test coverage for the following features while this PR is up and reviewable.

- [x] TODO: tests for new api-client/src/protocols/utils accessors
- [x] TODO: test for DeckThumbnail
- [x] TODO: test for ChooseRobotSlideout
- [x] TODO: test for ProtocolDetails organism
- [x] TODO: test for protocol-storage redux
- [x] TODO: test for config migration 

# Changelog

- include app-shell machinery to execute python
- include new config for specifying an alternative path at which to find a python interpreter (dev defaults the the virtualenv of the monorepo's `api` dir) 
- initiate local protocol analysis within app-shell protocol storage when protocol files are uploaded 
- build out analysis on protocols landing and detail pages. use shared protocol analysis file parsing utils exported from api-client

# Review requests

- Upload a protocol to the app, you should see it appear with parsed data after FE analysis succeeds.
- Navigate to that protocol's detail page.  
- Open the "Choose a Robot" slide-out from the "Run this protocol" CTA's in the overflow menu of the ProtocolCard on the landing page, as well as the Primary Button on the Protocol Detail page.

# Risk assessment
high, this wires up a lot of verticality across our stack.

### CODE REVIEW

File: app-shell/package.json [hunk 1]
```diff
@@ -55,7 +55,6 @@
     "node-fetch": "2.6.0",
     "node-stream-zip": "1.8.2",
     "pump": "3.0.0",
-    "python-shell": "^3.0.1",
```
Comment (b-cooper) [inline app-shell/package.json:58]: removed this dependency that I had added recently with hopes of using it for FE python execution, but is ultimately unnecessary

File: app-shell/src/python/index.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+export function runFileWithPython(
+  srcFilePath: string,
+  destFilePath: string
+): Promise<void> {
+  const pythonPath = findPython()
+
+  if (pythonPath != null) {
+    return execa(pythonPath, [
+      '-m',
+      'opentrons.cli.__init__',
```
Comment (mcous) [inline app-shell/src/python/index.ts]: With #9862 merged, you can drop the `__init__`

```suggestion
      'opentrons.cli',
```

File: app/src/atoms/Slideout/index.tsx [hunk 3]
```diff
@@ -100,13 +101,9 @@ export const Slideout = (props: Props): JSX.Element | null => {
               paddingX={SPACING_3}
```
Comment (koji) [inline app/src/atoms/Slideout/index.tsx]: maybe `SPACING.spacing4` instead of `SPACE_3`

File: app/src/organisms/ChooseRobotSlideout/index.tsx [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  getReachableRobots,
+  getScanning,
+  getUnreachableRobots,
+  startDiscovery,
+} from '../../redux/discovery'
+
+import { StyledText } from '../../atoms/text'
+import { StoredProtocolData } from '../../redux/protocol-storage'
+
+import { NavLink, useHistory } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/ChooseRobotSlideout/index.tsx]: this line should be after `react-redux` shouldn't it?

File: app/src/organisms/ChooseRobotSlideout/useCreateRunFromProtocol.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import { useQueryClient } from 'react-query'
+import {
+  useHost,
+  useCreateProtocolMutation,
+  useCreateRunMutation,
+} from '@opentrons/react-api-client'
+
+import type { UseMutateFunction } from 'react-query'
+import type { Protocol } from '@opentrons/api-client'
+import { UseCreateRunMutationOptions } from '@opentrons/react-api-client/src/runs/useCreateRunMutation'
```
Comment (koji) [inline app/src/organisms/ChooseRobotSlideout/useCreateRunFromProtocol.ts]: This line should be before `import type`?

File: app/src/organisms/ProtocolDetails/__tests__/ProtocolDetails.test.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+} from '../../../redux/discovery'
+import {
+  mockConnectableRobot,
+  mockReachableRobot,
+  mockUnreachableRobot,
+} from '../../../redux/discovery/__fixtures__'
+import { storedProtocolData } from '../../../redux/protocol-storage/__fixtures__'
+import { ProtocolDetails } from '..'
+import { DeckThumbnail } from '../../../molecules/DeckThumbnail'
+import { fireEvent } from '@testing-library/react'
```
Comment (koji) [inline app/src/organisms/ProtocolDetails/__tests__/ProtocolDetails.test.tsx]: Probably this line should be between `react-router-dom` and `i18`

File: app/src/organisms/ProtocolsLanding/EmptyStateLinks.tsx [hunk 7]
```diff
@@ -6,16 +6,14 @@ import {
   Text,
   Flex,
   FONT_SIZE_CAPTION,
-  SPACING_2,
-  SPACING_3,
+  SPACING,
   C_MED_GRAY,
   Link,
   SPACING_1,
```
Comment (koji) [inline app/src/organisms/ProtocolsLanding/EmptyStateLinks.tsx]: Probably removing this and replacing it with `SPACING.spacing2` would be good.
L51, L62, and L72

File: app/src/organisms/ProtocolsLanding/EmptyStateLinks.tsx [hunk 8]
```diff
@@ -33,12 +31,11 @@ export function EmptyStateLinks(props: Props): JSX.Element | null {
...
       alignItems={ALIGN_CENTER}
-      position={POSITION_ABSOLUTE}
-      bottom="0"
-      paddingBottom={SPACING_3}
+      marginY={SPACING.spacing6}
+      paddingBottom={SPACING.spacing3}
       width="96.5%"
     >
-      <Text role="complementary" as="h5" marginBottom={SPACING_2}>
+      <Text role="complementary" as="h5" marginBottom={SPACING.spacing2}>
```
Comment (koji) [inline app/src/organisms/ProtocolsLanding/EmptyStateLinks.tsx]:
```
export const SPACING_2 = '0.5rem'
```

```
export const spacing2 = '0.25rem' // 4px
export const spacing3 = '0.5rem' // 8px
```

maybe spacing3?

and using `StyledText` instead `Text` would be good.

File: app/src/organisms/ProtocolsLanding/ProtocolCard.tsx [hunk 9]
```diff
@@ -93,81 +64,111 @@ export function ProtocolCard(props: ProtocolCardProps): JSX.Element | null {
...
-          </Flex>
-        </Flex>
-
+        {mostRecentAnalysis != null ? (
+          <AnalysisInfo
+            mostRecentAnalysis={mostRecentAnalysis}
+            protocolDisplayName={protocolDisplayName}
+          />
+        ) : (
+          <StyledText>{t('loading_data')}</StyledText>
```
Comment (koji) [inline app/src/organisms/ProtocolsLanding/ProtocolCard.tsx]: `as="p"` ?

**Review by koji — COMMENTED** [review]
file-system.test.ts `writes a protocol file to a new directory` has an issue.

https://github.com/Opentrons/opentrons/runs/5837426226?check_suite_focus=true#step:8:140

**Comment by b-cooper** [comment]
> https://github.com/Opentrons/opentrons/runs/5837426226?check_suite_focus=true#step:8:140

Yeah this is a failure in CI only now. It won't pass until the follow up work to bundle standalone python into the electron build is completed.  I'll find a way to swallow this error for now.

File: api-client/src/protocols/utils.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  LoadLabwareRunTimeCommand,
+  LoadModuleRunTimeCommand,
+  LoadPipetteRunTimeCommand,
+} from '@opentrons/shared-data/protocol/types/schemaV6/command/setup'
+
+interface PipetteNamesByMount {
+  left: PipetteName | null
+  right: PipetteName | null
+}
+export function parseInitialPipetteNamesByMount(
```
Comment (brenthagen) [inline api-client/src/protocols/utils.ts:16]: perhaps not a priority at the moment, but there's similar parsing happening elsewhere, like in `useRunPipetteInfoByMount`, that would be nice to extract like this

**Review by brenthagen — APPROVED** [review]
super exciting!! ✨

**Pull request merged**